### ISSUE TITLE
proposal to improve genotype management interface

### ISSUE THREAD

**Issue by ValWood** [pombase/canto#1381]
We chatted about this yesterday.  Essentially there are a few problems we want to resolve.
1. We want to reduce the excessive amount of clicking required generally. Adding single alleles takes quite a few  (?) steps.
2. We want to include short cuts to create single gene delation alleles
3. We want to improve this bottleneck for users. We think it is possible a lot of people bail here, since we already know that some people realise they have done something wrong and just submit a partially complete session.

We came up with a slightly different approach to distinguish between single and multi gene phenotypes.
Basically the gene list will be on the left of the page, with the ability to create single allele genotypes directly from the list. 
All genotype will appear on the left of the page. 
Single gene genotypes will  appear on the right in their own table.
Multi gene alleles will be created from single gene alleles and will also appear in their own table, underneath single allele genotypes.

here is a mock-up
![untitled presentation 3](https://user-images.githubusercontent.com/7359272/36026879-d82ada4a-0d90-11e8-97c3-017beaa3cefd.jpg)

Comments everyone....

Outstanding:
What to do about background (I prefer to add this at annotation time, I think)
We also need to think about this in the contect of PHI-Canto development, but I think a similar 
set up will be useful for pathogen vs host geneotypes.

**Comment by CuzickA** [comment]
Mock up looks good. I'll have a think about it in the context of PHI-Canto. @ValWood did you see the PHI-Canto mock up idea I emailed earlier in the week? 
We will have 
pathogen gene information (uniprot), 
host gene information (uniprot) and 
host species name where there is no specific host gene of interest (ncbi taxon id). 
Displaying the strain information is important and then thinking about how to link the bits of information to create the pathogen-host metagenotype. 
Both the pathogen and the host would need the single and multiple allele options.

**Comment by Antonialock** [comment]
I think that on the left hand side, there can be two buttons (instead of tickboxes)
"create deletion" and "create [other] allele"

this is so that you can create a single deletion with one click 
if the "create [other] allele" is pressed then the normal popup for creating an allele should show.

I think I also prefer to add the background at the point of making the phenotype annotation (background might differ between different experiments even though the "alleles of interest" remain the same). 
Perhaps a separate text box at the point of adding conditions?

**Comment by mah11** [comment]
Bear in mind that if we add background details as part of annotation, Canto will have to be able to create new genotypes "on the fly". If the backgrounds are different, the genotypes are different.

**Comment by mah11** [comment]
> I think that on the left hand side, there can be two buttons (instead of tickboxes)
"create deletion" and "create [other] allele"

I agree; that's what we discussed and buttons will be much better. Tickboxes go only in the upper right quadrant, for selecting (already-entered) single allele genotypes to be combined into a multi-allele genotype.

**Comment by ValWood** [comment]
> if the "create [other] allele" is pressed then the normal popup for creating an allele should show.

I think there could be a pull down to select the allele type here, so that the pop up only needs to prompt for details of the specific allele type you chose (one less action)

**Comment by ValWood** [comment]
@CuzickA  i only had a quick glance,  I'm a bit behind this week.
I think these changes should work well with the  pathogen part to, and it will iron out some long-standing issues that our users have had, and some major frustrations for us (too much clicking).

**Comment by CuzickA** [comment]
No worries. Yes it would be nice to reduce the amount of clicking.

**Comment by ValWood** [comment]
Omit the identifier column to save screen width?
(we don't include the ID in annotation tables for instance, gene name is enough)

**Comment by ValWood** [comment]
I had a new one today!

![new one](https://user-images.githubusercontent.com/7359272/36112201-c38dc43e-1020-11e8-987e-0255cf0f3f15.jpg)

the histone phenotype is a mult-igene, but the ago1,chp1,clr4  was meant to be single alleles.

In  a novel approach the phenotype terms were added as allele names....

**Comment by kimrutherford** [comment]
I'm not sure I understand the new work flow.  Will the user need to create single alleles genotypes first, then combine them to make a multi-allele genotype?  Is that what the "Use in multi" label is about?

**Comment by ValWood** [comment]
yes that's correct.  We feel this would  be clearer for users, but also save lots of time creating multigene alleles with  slightly different allele combinations.

**Comment by kimrutherford** [comment]
> multigene alleles with slightly different allele combinations.

Do you mean multigene genotypes here?

> slightly different allele combinations

When you say allele here, do you mean allele+expression?  Or will the user set the expression of each allele when the multi-allele genotype is created?

Where does background fit in?  If a user creates a single allele genotype will they be able to set the background?  Will we only allow single allele genotypes to be combined into multi-allele genotypes if they all have the same background?

**Comment by ValWood** [comment]
allele+expression

this pop-up will still be used to add the details:
![tmp](https://user-images.githubusercontent.com/7359272/36964966-18ab42ae-2050-11e8-86ba-bb1ea459f8f1.jpg)

**Comment by ValWood** [comment]
(rather asimilar pop-up)

**Comment by ValWood** [comment]
Background
We need to discuss this.
Although formally the background is part of the genotype,  for our purposes it might be better to collect this on a per annotation basis.
Users largely want to see all of the same allele phenotypes together regardless of background, so perhaps when curating we don't want  every  genotype+background displayed in the genotype table ? 
Just a thought 

a) would keep the genotype interface much simpler
b) would make the background less likely to be overlooked, on a  per experiment basis.

(I  often find it confusing  when I'm creating mulit- alleles and my existing  single allele  has a background that isn't  used in later experiments, that I need to create another single gene allele)

this would not need to change the way the background was stored behind the scenes

**Comment by kimrutherford** [comment]
> this would not need to change the way the background was stored behind the scenes

I think it will need a database change if you want to add a background when you make an annotation.

**Comment by kimrutherford** [comment]
I've queried for genotypes with backgrounds in Chado and there are hardly any.  Is that expected?  Maybe I'm doing the wrong query?

I've put the list of the backgrounds and how many genotypes with that background in Dropbox:
Dropbox/pombase/Chado/queries/genotype_names-2018-03-12.txt
or:
https://www.dropbox.com/s/3vhqwpcwwomghvd/genotype_names-2018-03-12.txt?dl=0

**Comment by ValWood** [comment]
I have used a lot more than this?

**Comment by ValWood** [comment]
I updated the mock-up as discussed.
I removed any comments related to things which are now included in the mock-up
The background is still the issue  we will have a better idea when we see the list...

**Comment by ValWood** [comment]
- add examples where we use the background and how.
cdc22-block and release is a good example
- also  find some examples of where the same allele is needed with multiple different backgrounds (where is is a pain to create the different alleles, and where we think it would be easier to add the  background at annotation time

**Comment by kimrutherford** [comment]
> I have used a lot more than this?

Sorry about that.  I must have been querying a test database.  This list look better:
https://www.dropbox.com/s/2y93jeygn34tgwo/genotype_backgrounds-2018-03-16.txt?dl=0

**Comment by Antonialock** [comment]
looks like some of them should be the same but have typos

**Comment by ValWood** [comment]
@kimrutherford  could you run a version of this query and include canto session IDs?

I'd like to standardize a lot of them. I will do this on Monday morning if I get the list, because I want to move forward with a decision on how to handle background this week. I can find some good examples as I do this.

v

**Comment by ValWood** [comment]
I think we should standardize without commas? (we seem to exclude commas more than we include them)
Could you fix this part automagically? 
Then I will fix the rest manually.

**Comment by ValWood** [comment]
background fix was was requested back here too
https://github.com/pombase/canto/issues/1196

**Comment by kimrutherford** [comment]
> @kimrutherford could you run a version of this query and include canto session IDs?

In Dropbox/pombase/Chado/queries/genotype_backgrounds_and_sessions-2018-03-19.txt : 
  https://www.dropbox.com/s/3tcz1k8q06nsauj/genotype_backgrounds_and_sessions-2018-03-19.txt?dl=0

**Comment by ValWood** [comment]
I was thinking no commas, would make parsable (and is more common).
However, we have "homozygous diploid"
We do intend to something with these
https://github.com/pombase/canto/issues/976

@kimrutherford  what if for now, I put "double quotes" around any string with a space?
Would that work?

I will role the ticket above into the genotype proposal (summary, tick-box for ploidy).

**Comment by ValWood** [comment]
On second thoughts, maybe I should move to "other"    in the allele description. That is where the others are temporarily stashed. At least then, they are all in the same place?
@mah11 @Antonialock ?

**Comment by Antonialock** [comment]
I think that is how I do it already e.g.
ygf1-/yfg- homozygous diploid

though I might not be consistent for this with meiotic phenotypes. 
Heterozygous ones are really the ones that are interesting to differentiate

**Comment by Antonialock** [comment]
because the other one is a "deletion" anyway?

**Comment by kimrutherford** [comment]
Do we want genotype editing to work the same way as it does at the moment?

**Comment by ValWood** [comment]
I'm happy with how it works right now.

**Comment by kimrutherford** [comment]
Will it be OK if the workflow for creating genotypes is different from editing genotypes?  At the moment creating and editing are done with the same actions / same page.

**Comment by ValWood** [comment]
If it keeps the workflow simpler I would be happy for editing to be slightly different (it doesn't not need to happen so often hopefully)

**Comment by kimrutherford** [comment]
I've started implementing this.  I've mostly concentrated on simple things that don't require deep changes.  I'm completely ignoring genotype names and backgrounds until we finalise what to do.

I've released what I've done so far in the test Canto: https://curation.pombase.org/test

It's mostly layout changes so far.  I haven't implemented the action of combining single allele genotypes to get multi allele genotypes.

We can chat about it on the next Skype.

**Comment by mah11** [comment]
> I've released what I've done so far in the test Canto

The "Deletion" button is everything I'd hoped it would be :)

**Comment by Antonialock** [comment]
> The "Deletion" button is everything I'd hoped it would be :)

yep!

**Comment by ValWood** [comment]
I'm excited about this but haven't had time to test it yet...

**Comment by kimrutherford** [comment]
In the test Canto, I've added checkboxes beside the single allele genotypes and added a button to test combining single allele genotypes into multi-allele genotypes.  There are formatting and other problems but we can chat about those on the next Skype call.

**Comment by ValWood** [comment]
Very very nice..... 
I'm sure we can think of ways to  compact and format

here is a reference test session

https://curation.pombase.org/test/curs/2cc43eb21aa04239/genotype_manage#/select/11

![test eg](https://user-images.githubusercontent.com/7359272/38376152-de8a32fc-38ef-11e8-8205-0f2cf3616fcc.jpg)

**Comment by ValWood** [comment]
It is so, so, so much quicker and clearer this way.
- [ ] One minor suggestion, once you used "combine selected genotypes" the boxes should  "uncheck"

**Comment by ValWood** [comment]
I think the major formatting improvement would be to make  as much use as possible of the left /right real estate  for the allele tables....

**Comment by kimrutherford** [comment]
> I think the major formatting improvement would be to make as much use as possible of the
> left /right real estate for the allele tables....

Good idea.  I'll make the right head tables a bit wider.

I think it would be good to remove the genes column if possible.

**Comment by kimrutherford** [comment]
> One minor suggestion, once you used "combine selected genotypes" the boxes should "uncheck"

That's fixed now.

I've also made the genotypes tables on the right a bit wider.

**Comment by ValWood** [comment]
better...

for some reason the multi allele  blah MKGYP allele line wraps but the single allele doesn't.
It would be better to minimise the line wrapping in this filed by allowing it to be wider still...

![better](https://user-images.githubusercontent.com/7359272/38414358-2c20578a-3986-11e8-8cda-36e149d55f83.jpg)

**Comment by Antonialock** [comment]
fantastic! 

column header for left hand table: "create single allele genotype"

column header for tickbox column: "select for multi-allele genotype"? (a bit long)

remove right hand column from mult-allele table

increase space between action button and multi table
![screen shot 2018-04-06 at 10 43 52](https://user-images.githubusercontent.com/12727506/38414798-79a3f092-3987-11e8-93b7-a9ad7dbc88f5.png)

decrease space between lefthand column and right hand column

**Comment by ValWood** [comment]
These are not major space savers, but every little helps.

Suggestions:

- [ ]  place the tables a little higher
- [ ] Single allele genotypes (Mouse over genotypes for more actions) onto a  single line
- [ ] I think the "action" button to combine would be better  directly to the right of the single allele table instead of underneath

- [ ]  The expression section could be compacted if, at the point of selection you saw:

Overexpression  (OX)
Wild type product level  (WT)
Knockdown Null  (KD)
Not assayed (NA)
Null (Null)

 (or similar) and then show only abbreviations in the table)...Header could be Exp.

**Comment by Antonialock** [comment]
Val I edited your comment to change Overexpression  (OX) to 
Overexpression  (OP) (I think OP is more common, OX makes me think of oxidised)

**Comment by ValWood** [comment]
yes it is...in principle do you think abbreviation work OK....I can't  think why it wouldn't be if the abbreviations are clear at point of selection?

**Comment by Antonialock** [comment]
Yeah, I think it is fine, OP KD WT whould be straight forward. NA maybe not so much (looks like not applicable?....in the tool it is probably ok? or UNK for unknown)

**Comment by ValWood** [comment]
Antonia suggested

column header for tickbox column: "select for multi-allele genotype"? (a bit long)~

I  don't think a header is required if the instructions are clear, but in this case, I wonder if the tick boxes can be adjacent to the table rows (but outside the table).?

**Comment by kimrutherford** [comment]
I'm putting off these changes while I do some (hopefully invisible) HTML rearranging to help with the multi-organism / host pathogen requirements.

> Single allele genotypes (Mouse over genotypes for more actions) onto a single line

I don't understand this one.

> I think the "action" button to combine would be better directly to the right of the single allele table instead of underneath

To the right of the single allele table header?

**Comment by Antonialock** [comment]
> I think the "action" button to combine would be better directly to the right of the single allele table instead of underneath
> 
> To the right of the single allele table header?

I think it should at least be right justified

**Comment by ValWood** [comment]
https://github.com/pombase/canto/issues/1406
